src/osd/PGLog.cc

```diff
@@ -865,40 +872,39 @@ void PGLog::_write_log_and_missing(
   eversion_t write_from_dups,
   bool *may_include_deletes_in_missing_dirty, // in/out param
   set<string> *log_keys_debug,
+  set<string> *to_remove,
+  set<std::pair<string, string>> *to_rmkeyrange,
   const DoutPrefixProvider *dpp
   ) {
   ldpp_dout(dpp, 10) << __func__ << " clearing up to " << dirty_to
 		     << " dirty_to_dups=" << dirty_to_dups
 		     << " dirty_from_dups=" << dirty_from_dups
 		     << " write_from_dups=" << write_from_dups
 		     << " trimmed_dups.size()=" << trimmed_dups.size() << dendl;
-  set<string> to_remove;
-  to_remove.swap(trimmed_dups);
+  (*to_remove).swap(trimmed_dups);
```

Comment:
Here’s what I’m concerned about: Since `rmkeys` and `rmkeyrange` are related to `PGLog`, we need to track them in order to make those operations OP_LOG*. There are two ways to approach this: 1) (this PR) Add arguments for rmkeys and rmkeyrange to `_write_log_and_missing` and let `prepare_write` determine whether to use OP_LOG* or OP_OMAP*. 2) After `write_log_and_missing` without changes, have `prepare_write` go through the transaction to find `rmkeys` and `rmkeyrange`, then replace them with OP_LOG* if needed. Note that 2) doesn't require changes in PGLog.cc.

Reply:
@cyx1231st https://github.com/myoungwon/ceph/commit/5e80fa86e224cebef253b6be7a5ecf1d36931c32 This is the case 2 as explained above, which minimizes code changes on the classic path.